### DISCUSSION TITLE
What is the design philosophy of Kade Engine?

### DISCUSSION THREAD

**Discussion by EliteMasterEric** [KadeArchive/Kade-Engine#2043] [Q&A]
I have been working on my own mods for FNF utilizing Kade Engine as a base, and I have seen lots of areas where significant reworks to the codebase would be beneficial, for example with abstracting definitions of many aspects of the game to JSON files like characters or weeks, or moving certain behaviors to song events (where currently only BPM changes are managed), or commissioning someone to make and implement optional note splashes.

However, I want to confirm that these changes would be wanted before starting PRs for them. Is Kade Engine's focus on sticking closely to the vanilla codebase while performing minor tweaks, or does it allow for major reworks like some other engines have, which may cause difficulties when Week 8 comes out, for example.

Rather than getting permission for each individual change I would want to implement, I'm instead interested looking to know what the focus for the engine's continued development is, and get a general idea of what types of changes, in terms of scope, focus area, etc. would be easily accepted and which would be potentially declined as "out of scope".

**Comment by isakube** [comment]
> I have seen lots of areas where significant reworks to the codebase would be beneficial, for example with abstracting definitions of many aspects of the game to JSON files like characters or weeks, or moving certain behaviors to song events (where currently only BPM changes are managed)

The main belief behind this is that you will always be building from source, so there is no reason to make these a `json` for the sole purpose of making things easier. I don't think we're aiming for beginner-friendly asldkjfhasdlkfjhasdkljfhasdkjfhadlkj here, seeing as kade (i think) closed some PRs only made for accessibility

> However, I want to confirm that these changes would be wanted before starting PRs for them. Is Kade Engine's focus on sticking closely to the vanilla codebase while performing minor tweaks, or does it allow for major reworks like some other engines have, which may cause difficulties when Week 8 comes out, for example.

Can't answer. We've obviously done major reworks (charter, input system) but a lot of our additions are minor / medium tweaks.

> Rather than getting permission for each individual change I would want to implement, I'm instead interested looking to know what the focus for the engine's continued development is, and get a general idea of what types of changes, in terms of scope, focus area, etc. would be easily accepted and which would be potentially declined as "out of scope".

literally have no idea

```ini
PROKUBE
=====
Badges obtained:
[Stupid Badge]
```

**Reply by EliteMasterEric** [reply]
Thanks for the input.

> seeing as kade (i think) closed some PRs only made for accessibility

Any examples of this we can reference?

**Reply by EliteMasterEric** [reply]
> The main belief behind this is that you will always be building from source, so there is no reason to make these a json for the sole purpose of making things easier.

The vast majority of use cases don't demand much custom logic for the character animations. Despite this, each custom character has to be added to a massive case structure in the constructor, which is probably the worst factory method I have ever seen and is begging for a refactor.

**Reply by isakube** [reply]
> The vast majority of use cases don't demand much custom logic for the character animations. Despite this, each custom character has to be added to a massive case structure in the constructor, which is probably the worst factory method I have ever seen and is begging for a refactor.

it works though

> Any examples of this we can reference?

I seem to remember one but #895 is the only one i can think of, even tho Kade didnt close it. idk i just remember some accessibility pr getting closed for some reason.

**Reply by EliteMasterEric** [reply]
#895 appears to be one that the original author abandoned; my interpretation is that they personally felt their own code was bad and decided to remove it.

One good way to dig through it is to filter by [items with type PR, status closed, commented on by Kade](https://github.com/KadeDev/Kade-Engine/pulls?q=is%3Apr+is%3Aclosed+commenter%3AKadeDev).

As an aside, #895 looks like a good opportunity for me to pick up the change and make it suitable for general use.

**Comment by EliteMasterEric** [comment]
Here's an evaluation of the pull requests which Kade commented on and closed. I added ✔️ on features which are already implemented, ❌ for features which were rejected, and 💡 for ideas that could be made into a new PR.

* #1857: 💡 I wrote this PR. One of the changes (support for changing GF sprite in modcharts) was implemented with KE v1.7, and the others (support for changing health icons in modcharts) were made unusable by the changes. This feature currently doesn't exist and could be added, but I've postponed that because it's better to do changes like this in code, since modcharts aren't supported on all platforms. If modcharts become available on Mac and Linux builds (via a rework of lua_jit or by implementation of hscript modchart support) this feature could be useful. My goal with this change was to make a push towards allowing as much custom behavior as possible to be done in modcharts to avoid forcing recompilation.
* #1613: ✔️ Kade rewrote the feature separately.
* #1536: 💡 Kade wanted this feature implemented differently (press a key to skip cutscene rather than use options menu to skip cutscene).
* #1452: 💡 Lol this is literally the same change as #1536 implemented in the same problematic way.
* #1583: ❌ I don't know why this change from the vanilla game was chosen or what caused this to be rejected. No explanation was given. This issue is a key example for why this thread is necessary. 
* #1463: ❌/💡 This is a big one for understanding the philosophy of the engine. Mommy Mearest's `Mom Pose Left` pose in her XML is actually her right pose, and `Mom Left Pose` is the actual left pose, which is confusing for those editing the XML. However, since this issue exists in [the base game as well](https://github.com/ninjamuffin99/Funkin/blob/master/assets/week4/images/Mom_Assets.xml), any [XXX over Mom](https://gamebanana.com/mods/cats/7870) skins also use this naming convention and would break if installed into a version of Kade Engine that implemented this PR. It is clear from Kade's response that compatibility with existing Skin mods for Vanilla FnF is important.
    * Added bulb because technically, we can create a smart Character sprite handler; if [the sprite atlas contains](https://github.com/HaxeFlixel/flixel/blob/27960e3b66c8ebad0052973831960761971168fd/flixel/animation/FlxAnimationController.hx#L720) `Mom Right Pose`, use that, otherwise fall back to `Mom Pose Left`.
* #1353: ❌ Not sure what this change resolves nor why Kade rejected it.
* #1352: 💡 Button to skip the intro of a song. Rejected for having a messy implementation that creates a new file for every song it's used in.
* #1154: 💡 Adds an option where you can't die but scores are not saved. Rejected for having poor implementation.
    * Note sure what CH means in Kade's response.
* #1113: 💡 Adds an option to disable the record scratch when missing a note. Apparently this got rejected because... Kade was redoing the Options menu around this time? Definitely could be re-implemented.
* #1100: ❌ This line throws an error if the WebM extension library isn't built correctly on your machine.
* #1065: ❌ This is a really shitty PR it includes so many extra formatting changes.
* #1027: ✔️ Change was already implemented.
* #1006: ❌ / 💡 This change gets rid of the caching process with the load screen, and instead makes caching occur in parallel while the user is on the title screen. This is actually a pretty neat idea, but it got rejected. Not sure as to the philosophy of this, another reason this thread is needed.
* #932: ❌ Week 7 increased the amount that the HP icons move when bopping on the beat. Kade rejected this claiming this change "made it look like crap".
* #920: ❌/ 💡 Change was rejected due to conflicts with master, and is probably obsolete due to the major revamp that modcharting had.
  * TODO: Check if these changes are redundant or could be re-implemented with a new PR.
* #732: ❌ Another shitty PR that changes a bunch of random files.
* #487: ❌ Literally rejected for "not being [Kade's] cup of tea."
* #486: ❌ This one just... 
* #1978: ❌/ 💡 Lol this one is the funniest PR ever it literally just removes a feature entirely. 
  * I added a 💡 to this one because you could add `|| FlxG.save.data.showNoteTimings` to [this line](https://github.com/KadeDev/Kade-Engine/blob/stable/source/PlayState.hx#L3621) and add an Option to the options menu for those who don't want to see it.
* #1265: ✔️ I one day aspire to make a commit so good that [Kade calls me Babe](https://github.com/KadeDev/Kade-Engine/pull/1265#issuecomment-877871417)

Here are my notes on getting PRs approved:
* Don't make shitty PRs. Make changes small and concise, and remove unnecessary and unrelated changes such as whitespace changes and unrelated fixes. These hinder the review process.
  * When I open the PR and go to the Files Changed tab, all the lines of code that changed should be related to the change.
  * This is just good advice in general, check out a guide like this one: https://betterprogramming.pub/how-to-make-a-perfect-pull-request-3578fb4c112
* Think through what the best means of implementing a feature would be for a good user experience.
  * The Options menu is an easy place to put stuff but it's not the best place to put everything.
* Don't break compatibility with mods and skins people would want to put over Kade.
  * This means that if changing a file would break mods that replace it (like song replacements or character skins) you should either avoid those changes or put extra effort into making them backwards compatible.
* Don't piss off Kade I guess?
  * Kade just doesn't like some features I guess.

**Reply by isakube** [reply]
> Here's an evaluation of the pull requests which Kade commented on and closed. I added ✔️ on features which are already implemented, ❌ for features which were rejected, and 💡 for ideas that could be made into a new PR.

I did not realize you ever posted this, WOW

> since modcharts aren't supported on all platforms. If modcharts become available on Mac and Linux builds

They are now! (You know this already though)

> Not sure what Kade means by CH in his response

Clone Hero. In Clone Hero's practice mode, you can choose sections of the song to play (Chorus, Verse, Bridge, etc.) or all of it at once from a certain point.

> Don't piss off Kade I guess?

Health percentage :trollface:
